src/butterworth_filter.cpp

```diff
@@ -0,0 +1,92 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2022, PickNik Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Andy Zelenak
+   Description: A first-order Butterworth low-pass filter. There is only one parameter to tune.
+ */
+
+#include "trackjoint/butterworth_filter.h"
+
+namespace trackjoint
+{
+namespace
+{
+constexpr double EPSILON = 1e-9;
+}
+
+ButterworthFilter::ButterworthFilter(double low_pass_filter_coeff)
+  : previous_measurements_{ 0., 0. }
+  , previous_filtered_measurement_(0.)
+  , scale_term_(1. / (1. + low_pass_filter_coeff))
+  , feedback_term_(1. - low_pass_filter_coeff)
```

Comment:
Is this supposed to be (1 - low_pass_filter_coeff) / (1 + low_pass_filter_coeff)?

In Octave, `butter(1, 0.2)` gives:
![image](https://user-images.githubusercontent.com/10466537/151384839-9a3436f7-6786-4a8e-83bb-f5b58fff67ee.png)
![image](https://user-images.githubusercontent.com/10466537/151384916-3f9dbbfe-933c-4924-abcb-ecd73e6eeaa4.png)
![image](https://user-images.githubusercontent.com/10466537/151384995-d69fce3a-3825-485d-9f20-f4d565679c58.png)
![image](https://user-images.githubusercontent.com/10466537/151385087-06e496e5-353e-4f25-bb83-5dd7c1ab37e3.png)

This corresponds to a Butterworth filter of order 1, with a cutoff frequency of 0.1 * sampling frequency (1 corresponds to half the sampling frequency).

In this case, with `low_pass_coeff` c:
![image](https://user-images.githubusercontent.com/10466537/151387268-67daa5e2-7bd8-4498-a72e-9225fe1fdd5b.png)

we get a corresponding c of:
![image](https://user-images.githubusercontent.com/10466537/151387480-f28aa14c-46de-42da-8389-35ebfbef5a82.png)

or:
![image](https://user-images.githubusercontent.com/10466537/151387627-6c31c004-2b7f-49eb-a1b4-608492dbb43c.png)

If we say for `scale term`:
![image](https://user-images.githubusercontent.com/10466537/151387982-dccbd0e2-db8e-44d6-ade9-5f982c014b26.png)

using the above c:
![image](https://user-images.githubusercontent.com/10466537/151384839-9a3436f7-6786-4a8e-83bb-f5b58fff67ee.png)

This matches with the coefficient from `butter`. Same follows with b_1.

However, for `feedback_term`, using
![image](https://user-images.githubusercontent.com/10466537/151388544-26e764b8-d755-4524-8698-ff5e4fec91f8.png)

and substituting in our c value gives us:
![image](https://user-images.githubusercontent.com/10466537/151388672-801cc85f-5e6b-4880-87b1-6ecda1d69eb3.png)

which doesn't match the above output from `butter`. However, using:
![image](https://user-images.githubusercontent.com/10466537/151388966-98cafdd4-800e-4604-a0cd-25fd3589bb49.png)

and substituting in our c value from earlier, we get:
![image](https://user-images.githubusercontent.com/10466537/151385087-06e496e5-353e-4f25-bb83-5dd7c1ab37e3.png)

which matches the `butter` output.

I have some scribbles in a notebook that I can show you if you want to see more detail.